src/org/traccar/smpp/SmppClient.java

```diff
@@ -0,0 +1,218 @@
+                thread.setName("Reconnection-" + name);
+                return thread;
+            }
+        });
+
+        scheduleReconnect();
+    }
+
+    public synchronized SmppSession getSession() {
+        return smppSession;
+    }
+
+    public String mapDataCodingToCharset(byte dataCoding) {
+        switch (dataCoding) {
+            case SmppConstants.DATA_CODING_LATIN1:
+                return CharsetUtil.NAME_ISO_8859_1;
+            case SmppConstants.DATA_CODING_UCS2:
+                return CharsetUtil.NAME_UCS_2;
+            default:
+                return CharsetUtil.NAME_GSM;
+        }
+    }
+
+    protected synchronized void reconnect() {
+        try {
+            disconnect();
+            smppSession = clientBootstrap.bind(sessionConfig, sessionHandler);
+            stopReconnectionkTask();
+            runEnquireLinkTask();
+            Log.info("Smpp session connected");
+        } catch (SmppTimeoutException | SmppChannelException
+                | UnrecoverablePduException | InterruptedException error) {
+            Log.warning("Unable to connect to smpp server: ", error);
+        }
+    }
+
+    public void scheduleReconnect() {
+        reconnectionTask = reconnectionExecutor.scheduleWithFixedDelay(
+                new ReconnectionTask(this),
+                reconnectionDelay, reconnectionDelay, TimeUnit.MILLISECONDS);
+    }
+
+    private void stopReconnectionkTask() {
+        if (reconnectionTask != null) {
+            reconnectionTask.cancel(false);
+        }
+    }
+
+    private void disconnect() {
+        stopEnquireLinkTask();
+        destroySession();
+    }
+
+    private void runEnquireLinkTask() {
+        enquireLinkTask = enquireLinkExecutor.scheduleWithFixedDelay(
+                new EnquireLinkTask(this, enquireLinkTimeout),
+                enquireLinkPeriod, enquireLinkPeriod, TimeUnit.MILLISECONDS);
+    }
+
+    private void stopEnquireLinkTask() {
+        if (enquireLinkTask != null) {
+            enquireLinkTask.cancel(true);
+        }
+    }
+
+    private void destroySession() {
+        if (smppSession != null) {
+            Log.debug("Cleaning up smpp session... ");
+            smppSession.destroy();
+            smppSession = null;
+        }
+    }
+
+    public synchronized void sendMessageSync(String destAddress, String message) throws RecoverablePduException,
+            UnrecoverablePduException, SmppTimeoutException, SmppChannelException, InterruptedException {
+        if (getSession() != null && getSession().isBound()) {
+            byte[] textBytes = CharsetUtil.encode(message, charsetName);
+
+            SubmitSm submit = new SubmitSm();
+            submit.setSourceAddress(new Address(sourceTon, sourceNpi, sourceAddress));
+            submit.setDestAddress(new Address(destTon, destNpi, destAddress));
+            submit.setRegisteredDelivery(SmppConstants.REGISTERED_DELIVERY_SMSC_RECEIPT_REQUESTED);
```

Comment:
Not really for now. Right, just log.
I've made it more for testing. Will remove it now. We can revert it in future, when will control delivery.